### PR TITLE
various fixes for schemes

### PR DESCRIPTION
**Pull request by simonbrandhorst**
Various fixes and preparations for elliptic surfaces.

### CODE REVIEW

File: experimental/Schemes/Auxiliary.jl [hunk 1]
```diff
@@ -126,6 +126,16 @@ function pullback(f::AbsCoveredSchemeMorphism, C::EffectiveCartierDivisor)
   return EffectiveCartierDivisor(X, triv_dict, trivializing_covering=triv_cov, check=false)
 end
 
+function pullback(f::AbsCoveredSchemeMorphism, C::CartierDivisor)
```
Comment (HechtiDerLachs) [inline experimental/Schemes/Auxiliary.jl:129]: Had this only been done for effective cartier divisors?

File: experimental/Schemes/WeilDivisor.jl [hunk 2]
```diff
@@ -38,6 +38,7 @@ stored as a formal linear combination over some ring ``R`` of
         isprime(D) || error("components of a divisor must be sheaves of prime ideals")
         dim(X) - dim(D) == 1 || error("components of a divisor must be of codimension one")
       end
+      true # Return true here for the check macro as all internal checks have passed.
```
Comment (HechtiDerLachs) [inline experimental/Schemes/WeilDivisor.jl]:
```suggestion
```

File: experimental/Schemes/WeilDivisor.jl [hunk 3]
```diff
@@ -49,6 +50,7 @@ stored as a formal linear combination over some ring ``R`` of
...
```
Comment (HechtiDerLachs) [inline experimental/Schemes/WeilDivisor.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl [hunk 4]
```diff
@@ -51,6 +51,7 @@ function restrict(f::SpecMor, U::SpecOpen, V::SpecOpen; check::Bool=true)
     issubset(U, domain(f)) || error("$U is not contained in the domain of $f")
     issubset(V, codomain(f)) || error("$V is not contained in the codomain of $f")
     all(x->issubset(preimage(f, x), U), affine_patches(V)) || error("preimage of $V is not contained in $U")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl [hunk 5]
```diff
@@ -65,6 +66,7 @@ function restrict(
     issubset(U, domain(f)) || error("the given open is not an open subset of the domain of the map")
     issubset(V, codomain(f)) || error("the given open is not an open subset of the codomain of the map")
     issubset(preimage(f,V), U) || error("f(U) is not contained in V")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]:
```suggestion
```
These are SO annoying!!! Why do we keep on picking them up all the time???

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl [hunk 6]
```diff
@@ -77,6 +79,7 @@ function restrict(f::SpecOpenMor, W::AbsSpec, Y::AbsSpec; check::Bool=true)
   @check begin
     issubset(W, domain(f)) || error("$U is not contained in the domain of $f")
     issubset(W, preimage(f, Y)) || error("image of $W is not contained in $Y")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Types.jl [hunk 7]
```diff
@@ -51,6 +51,7 @@ mutable struct SpecOpenMor{DomainType<:SpecOpen,
       for g in f
         is_empty(subscheme(domain(g), pullback(g).(complement_equations(V)))) || error("image is not contained in the codomain")
       end
+      true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Types.jl]:
```suggestion
```

Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Types.jl]: ...and so on below. I'm getting tired of this, sorry.

File: src/Rings/mpolyquo-localizations.jl [hunk 8]
```diff
@@ -1054,8 +1056,7 @@ end
 # uncommon implementation of the numerator and denominator methods
 function (f::MPolyQuoLocalizedRingHom)(a::AbsLocalizedRingElem)
   parent(a) === domain(f) || return f(domain(f)(a))
-  isone(lifted_denominator(a)) && return codomain(f)(restricted_map(f)(lifted_numerator(a)))
-  return codomain(f)(restricted_map(f)(lifted_numerator(a)))*inv(codomain(f)(restricted_map(f)(lifted_denominator(a))))
+  return codomain(f)(restricted_map(f)(lifted_numerator(a)), check=false)*inv(codomain(f)(restricted_map(f)(lifted_denominator(a)), check=false))
```
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl]: Why take this back?

**Review by HechtiDerLachs — COMMENTED** [review]
Looks good apart from those very annoying merge artifacts which pop up all the time.

File: experimental/Schemes/Auxiliary.jl (hunk 1 above)
Comment (simonbrandhorst) [reply experimental/Schemes/Auxiliary.jl:129]: I think so.

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl (hunk 5 above)
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]: Sorry, I thought they were necessary. So I included them. I can remove them.

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 9]
```diff
@@ -215,6 +215,7 @@ function restriction_map(
   @check begin
     X == hypersurface_complement(Y, h) || error("$X is not the hypersurface complement of $h in the ambient variety of $U")
     issubset(X, U) || error("$X is not a subset of $U")
+    true
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion

```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 10]
```diff
@@ -305,6 +306,7 @@ function restriction_map(U::SpecOpen{<:AbsSpec{<:Ring, <:AbsLocalizedRing}},
   @check begin
     issubset(X, Y) || error("$X is not contained in the ambient scheme of $U")
     issubset(X, U) || error("$X is not a subset of $U")
+    true
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 11]
```diff
@@ -329,6 +331,7 @@ function restriction_map(U::SpecOpen{<:AbsSpec{<:Ring, <:MPolyQuoRing}},
...
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 12]
```diff
@@ -344,6 +347,7 @@ function restriction_map(U::SpecOpen{<:AbsSpec{<:Ring, <:MPolyRing}},
...
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 13]
```diff
@@ -420,6 +424,7 @@ function canonical_isomorphism(S::SpecOpenRing, T::SpecOpenRing; check::Bool=tru
   R == ambient_coordinate_ring(Y) || error("rings can not be canonically compared")
   @check begin
     (domain(S) == domain(T)) || error("open domains are not isomorphic")
+    true
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/Rings/MPolyMap/MPolyQuo.jl [hunk 14]
```diff
@@ -83,6 +83,7 @@ function hom(R::MPolyQuoRing, S::NCRing, coeff_map, images::Vector; check::Bool
   @check begin
     _check_imgs_quo(R, S, imgs, coeff_map)
     _check_homo(S, imgs)
+    true
```
Comment (simonbrandhorst) [inline src/Rings/MPolyMap/MPolyQuo.jl]:
```suggestion
```

File: src/Rings/MPolyMap/MPolyRing.jl [hunk 15]
```diff
@@ -107,6 +107,7 @@ function hom(R::MPolyRing, S::NCRing, coeff_map, images::Vector; check::Bool = t
   @check begin
     _check_imgs(S, imgs)
     _check_homo(S, imgs) # defined in MPolyAnyMap.jl
+    true
```
Comment (simonbrandhorst) [inline src/Rings/MPolyMap/MPolyRing.jl]:
```suggestion
```

File: src/Rings/MPolyMap/MPolyRing.jl [hunk 16]
```diff
@@ -119,6 +120,7 @@ function hom(R::MPolyRing, S::NCRing, images::Vector; check::Bool = true)
...
```
Comment (simonbrandhorst) [inline src/Rings/MPolyMap/MPolyRing.jl]:
```suggestion
```

File: src/Rings/mpolyquo-localizations.jl (hunk 8 above)
Comment (simonbrandhorst) [reply src/Rings/mpolyquo-localizations.jl]: No clue, I did not know what was the old version and what the new one.
So I will revert to the old version.

Comment (simonbrandhorst) [inline src/Rings/mpolyquo-localizations.jl]:
```suggestion
  isone(lifted_denominator(a)) && return codomain(f)(restricted_map(f)(lifted_numerator(a)))
  codomain(f)(restricted_map(f)(lifted_numerator(a)))*inv(codomain(f)(restricted_map(f)(lifted_denominator(a))))
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl (hunk 5 above)
Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]: Oh! Sorry, I didn't know that. I used to get these `true` statements from faulty merges and had to remove them again and again all over the place. My previous comment was not intended against your work!

File: src/Rings/mpolyquo-localizations.jl [hunk 17]
```diff
@@ -1055,7 +1057,7 @@ end
 function (f::MPolyQuoLocalizedRingHom)(a::AbsLocalizedRingElem)
   parent(a) === domain(f) || return f(domain(f)(a))
   isone(lifted_denominator(a)) && return codomain(f)(restricted_map(f)(lifted_numerator(a)))
-  return codomain(f)(restricted_map(f)(lifted_numerator(a)))*inv(codomain(f)(restricted_map(f)(lifted_denominator(a))))
+  codomain(f)(restricted_map(f)(lifted_numerator(a)))*inv(codomain(f)(restricted_map(f)(lifted_denominator(a))))
```
Comment (simonbrandhorst) [inline src/Rings/mpolyquo-localizations.jl]:
```suggestion
  return codomain(f)(restricted_map(f)(lifted_numerator(a)))*inv(codomain(f)(restricted_map(f)(lifted_denominator(a))))
```

**Comment by simonbrandhorst** [comment]
Good to go from my part. I set to auto merge.

File: experimental/Schemes/IdealSheaves.jl [hunk 18]
```diff
@@ -444,9 +444,31 @@ function isone(I::IdealSheaf)
...
+    is_prime(I::IdealSheaf) -> Bool
+
+Return whether ``I`` is prime.
+
+We say that a sheaf of ideals is prime if its support is irreducible and
+``I`` is locally prime. (Note that the empty set is not irreducible.)
+"""
+function is_prime(I::IdealSheaf)
+  is_locally_prime(I) || return false
+  # TODO: this can be made more efficient
```
Comment (afkafkafk13) [inline experimental/Schemes/IdealSheaves.jl:457]: Please add a TODO that we have to make sure that we are on a connected scheme.

**Review by afkafkafk13 — APPROVED** [review]
Looks good to me.

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl [hunk 19]
```diff
@@ -258,6 +258,8 @@ julia> standard_spec(X)
 Spec of Localization of quotient of multivariate polynomial ring at complement of prime ideal
 ```
 """
+standard_spec(X::AbsSpec{<:Any, <:MPolyLocRing}) = Spec(MPolyQuoLocRing(ambient_coordinate_ring(X), ideal(ambient_coordinate_ring(X), [zero(ambient_coordinate_ring(X))]), inverted_set(OO(X))))
+
```
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl:262]: This looks like a mistake (perhaps made during merging), as the exact same method is defined in line 236. As a result we now get these warnings:
```
WARNING: Method definition standard_spec(Oscar.AbsSpec{var"#s1896", var"#s1895"} where var"#s1895"<:(Oscar.MPolyLocRing{BaseRingType, BaseRingElemType, RingType, RingElemType, MultSetType} where MultSetType<:Oscar.AbsMPolyMultSet{BaseRingType, BaseRingElemType, RingType, RingElemType} where RingElemType where RingType where BaseRingElemType where BaseRingType) where var"#s1896") in module Oscar at /home/runner/work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl:236 overwritten at /home/runner/work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl:261.
```
@HechtiDerLachs @simonbrandhorst please fix

Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl:262]: will do thx

Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl:262]: #2512

**Pull request merged**